arroyo/processing/strategies/routing_producer.py

```diff
@@ -0,0 +1,156 @@
+    ) -> None:
+        self.__commit_function = commit_function
+        self.__message_router = message_router
+        self.__closed = False
+        self.__offsets_to_be_committed: MutableMapping[Partition, Position] = {}
+        self.__queue: Deque[
+            Tuple[Message[KafkaPayload], Future[Message[KafkaPayload]]]
+        ] = deque()
+
+    def poll(self) -> None:
+        """
+        Periodically check which messages have been produced successfully and
+        call the commit function for the offsets that have been processed.
+        In order to commit offsets we need strict ordering, hence we stop
+        processing the queue of pending futures as soon as we encounter one
+        which is not yet completed.
+        """
+        while self.__queue:
+            message, future = self.__queue[0]
+
+            if not future.done():
+                break
+
+            exc = future.exception()
+            if exc is not None:
+                raise exc
+
+            self.__queue.popleft()
+            self.__offsets_to_be_committed[message.partition] = Position(
+                message.next_offset, message.timestamp
+            )
+
+        self.__commit_function(self.__offsets_to_be_committed)
+
+    def submit(self, message: Message[KafkaPayload]) -> None:
+        """
+        This is where the actual routing happens. Each message is passed to the
+        message router to get the producer and topic to which the message should
+        be sent. The message is then sent to the producer and the future is
+        stored in the queue.
+        """
+        assert not self.__closed
+        producer, topic = self.__message_router.get_route_for_message(message)
+        self.__queue.append(
+            (message, producer.produce(destination=topic, payload=message.payload))
+        )
+
+    def terminate(self) -> None:
+        self.__closed = True
+
+    def close(self) -> None:
+        self.__closed = True
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        """
+        In addition to flushing the queue, this method also calls the
+        shutdown of the router.
+        """
+        start = time.time()
+
+        # Commit all previously staged offsets
+        self.__commit_function({}, force=True)
+
+        while self.__queue:
+            remaining = timeout - (time.time() - start) if timeout is not None else None
+            if remaining is not None and remaining <= 0:
+                logger.warning(f"Timed out with {len(self.__queue)} futures in queue")
+                break
+
+            message, future = self.__queue.popleft()
+            future.result(remaining)
+            offset = {message.partition: message.position_to_commit}
+
+            logger.info("Committing offset: %r", offset)
+            self.__commit_function(offset, force=True)
+
+        remaining = timeout - (time.time() - start) if timeout is not None else None
+        self.__message_router.shutdown(remaining)
```

Comment:
You are still shutting down the message router here though (which I assume generally will be the one closing the producers). I don't think we should do this at all as they are not being created by this strategy. I think that whoever creates them should be the one calling close/shutdown on them.

Reply:
I would disagree. If you look at the shutdown sequence, there is no way for the creator of the KafkaProducer to perform a clean close. A shutdown would trigger a clean close() of the consumer here. If you look at how a strategy is closed https://github.com/getsentry/arroyo/blob/main/arroyo/processing/processor.py#L61-L78, it just calls the close() and join() methods of the strategies. The processor factory would create the KafkaProducers here https://github.com/getsentry/arroyo/blob/main/arroyo/processing/processor.py#L39. But once they are created the control is lost.

Lets take this example https://github.com/getsentry/arroyo/blob/main/examples/transform_and_produce/__init__.py#L27-L52. The producer for this gets called from https://github.com/getsentry/arroyo/blob/main/examples/transform_and_produce/script.py#L24-L29. And the last step is processor.run(). All signal handling is performed by the processor. Where would the close for this producer happen in your opinion?

Reply:
Yes, the example is missing the close. I think closing should be done in the signal handler e.g. https://github.com/getsentry/snuba/blob/master/snuba/cli/subscriptions_executor.py#L162-L166. Otherwise producer will be closed after a rebalance and stop working after that.

Reply:
Since this is a blocker, rather than going back and forth on this, I am going to build this in Sentry to make progress on the project. I also feel like seeing the implementation end to end would provide more context and help clarify things much more easily. Going to close this PR.